rclpy/src/rclpy/_rclpy.c

```diff
@@ -3446,10 +3446,12 @@ rclpy_message_info_to_dict(rmw_message_info_t * message_info)
   PyObject * source_timestamp = PyLong_FromLongLong(message_info->source_timestamp);
   if (source_timestamp != NULL) {
     PyDict_SetItemString(dict, "source_timestamp", source_timestamp);
+    Py_DecRef(source_timestamp);
```

Comment:
I'll preface this comment by saying that I don't fully understand the reference counting mechanism in Python.  So this probably requires someone else to look at it.

But I'm not convinced this is the right thing to do.  The [documentation](https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItemString) for `PyDict_SetItemString` says this:

```
This function does not steal a reference to val.
```

If `PyDict_SetItemString` doesn't steal a reference, then by doing a decref here, aren't there now *zero* references to the data?

Reply:
IIUC "stealing" in the CPython API means the function does not increase the reference count of the object passed to it. If you bake a cake you own it :cake: . If I steal the cake from you, you don't get to eat it, I do :yum: . When a function "steals" your reference, you don't get to decref it because it's not yours anymore.

`PyLong_FromLongLong()` creates a new py object with a reference count of = 1. `PyDict_SetItemString()` does not steal the reference, so it must increase the reference count to 2. This `Py_DecRef` brings the reference count back to 1; meaning the dictionary is the only one referencing the object.

Reply:
> `PyLong_FromLongLong()` creates a new py object with a reference count of = 1.

Right.

> `PyDict_SetItemString()` does not steal the reference, so it must increase the reference count to 2.

This is the part I wasn't clear on (and the documentation didn't clarify).  In the CPython API, is it just assumed that whenever you pass a `PyObject` into a function, a new reference is taken?

Reply:
> whenever you pass a PyObject into a function, a new reference is taken?

if it does not steal, i think yes.